quantum/lamparray.c

```diff
@@ -0,0 +1,171 @@
+    rgb_matrix_overlay_enable(enable);
+}
+
+void lamparray_backing_set_item(uint16_t index, lamp_state_t color) {
+    rgb_matrix_overlay_set(index, (rgba_t){color.red, color.green, color.blue, color.intensity ? 0 : 0xFF});
+}
+
+void lamparray_backing_update_finished(void) {
+    rgb_matrix_overlay_flush();
+}
+#endif
+
+static uint16_t cur_lamp_id   = 0;
+static bool     is_autonomous = true;
+
+void lamparray_get_attributes(lamparray_attributes_t* data) {
+    data->lamp_count      = LAMPARRAY_LAMP_COUNT;
+    data->update_interval = LAMPARRAY_UPDATE_INTERVAL;
+    data->kind            = LAMPARRAY_KIND_KEYBOARD;
+    data->bounds.width    = LAMPARRAY_WIDTH;
+    data->bounds.height   = LAMPARRAY_HEIGHT;
+    data->bounds.depth    = LAMPARRAY_DEPTH;
+}
+
+void lamparray_get_lamp_info(uint16_t lamp_id, lamparray_attributes_response_t* data) {
+    data->lamp_id         = lamp_id;
+    data->update_latency  = 1000;
+    data->is_programmable = 1;
+    data->input_binding   = lamparray_get_lamp_binding(lamp_id);
+
+    data->levels.red       = LAMPARRAY_RED_LEVELS;
+    data->levels.green     = LAMPARRAY_GREEN_LEVELS;
+    data->levels.blue      = LAMPARRAY_BLUE_LEVELS;
+    data->levels.intensity = LAMPARRAY_INTENSITY_LEVELS;
+
+    lamparray_get_lamp_info_data(lamp_id, data);
+}
+
+void lamparray_get_attributes_response(lamparray_attributes_response_t* data) {
+    lamparray_get_lamp_info(cur_lamp_id, data);
+
+    // Automatic address pointer incrementing - 26.8.1 LampAttributesRequestReport
+    cur_lamp_id++;
+    if (cur_lamp_id >= LAMPARRAY_LAMP_COUNT) cur_lamp_id = 0;
+}
+
+void lamparray_set_attributes_response(uint16_t lamp_id) {
+    cur_lamp_id = lamp_id;
+}
+
+void lamparray_set_control_response(uint8_t autonomous) {
+    is_autonomous = !!autonomous;
+
+    lamparray_backing_enable(!autonomous);
+}
+
+void lamparray_set_range(lamparray_range_update_t* data) {
+    // Any Lamp*UpdateReports can be ignored - 26.10.1 AutonomousMode
+    if (is_autonomous) {
+        return;
+    }
+
+    // Ensure IDs are within bounds
+    if ((data->start >= LAMPARRAY_LAMP_COUNT) || (data->end >= LAMPARRAY_LAMP_COUNT)) {
+        return;
+    }
+
+    for (uint16_t index = data->start; index <= data->end; index++) {
+        lamparray_backing_set_item(index, data->color);
+    }
+
+    // Batch update complete - 26.11 Updating Lamp State
+    if (data->flags & LAMP_UPDATE_FLAG_COMPLETE) {
+        lamparray_backing_update_finished();
+    }
+}
+
+void lamparray_set_items(lamparray_multi_update_t* data) {
+    // Any Lamp*UpdateReports can be ignored - 26.10.1 AutonomousMode
+    if (is_autonomous) {
+        return;
+    }
+
+    // Ensure data is within bounds
+    if (data->count > LAMP_MULTI_UPDATE_LAMP_COUNT) {
+        return;
+    }
+
+    // Limit iteration so IDs are within bounds
+    for (uint8_t i = 0; i < MIN(data->count, LAMPARRAY_LAMP_COUNT); i++) {
+        lamparray_backing_set_item(data->ids[i], data->colors[i]);
```

Comment:
```suggestion
        if (data->ids[i] < LAMPARRAY_LAMP_COUNT) {
            lamparray_backing_set_item(data->ids[i], data->colors[i]);
        }
```

Reply:
Even that is not 100% correct, because apparently the standard requires the device to ignore the whole report if any single value is incorrect:
![HID Usage Tables 1.4](https://github.com/qmk/qmk_firmware/assets/616082/7ea01ccd-844c-413f-87d5-5cbb2e35c5a1)
Not sure whether performing all those checks would be really good though (especially the duplicate ID check would be annoying to implement).